extensions/couchbase/controllerservices/CouchbaseClusterService.cpp

```diff
@@ -0,0 +1,167 @@
+/**
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "CouchbaseClusterService.h"
+#include "couchbase/codec/raw_binary_transcoder.hxx"
+#include "couchbase/codec/raw_string_transcoder.hxx"
+#include "couchbase/codec/raw_json_transcoder.hxx"
+
+#include "core/Resource.h"
+
+namespace org::apache::nifi::minifi::couchbase {
+
+namespace {
+
+constexpr std::array<::couchbase::errc::common, 9> temporary_connection_errors = {
+  ::couchbase::errc::common::temporary_failure,
+  ::couchbase::errc::common::request_canceled,
+  ::couchbase::errc::common::internal_server_failure,
+  ::couchbase::errc::common::cas_mismatch,
+  ::couchbase::errc::common::ambiguous_timeout,
+  ::couchbase::errc::common::unambiguous_timeout,
+  ::couchbase::errc::common::rate_limited,
+  ::couchbase::errc::common::quota_limited
+};
+
+CouchbaseErrorType getErrorType(const std::error_code& error_code) {
+  for (const auto& temporary_error : temporary_connection_errors) {
+    if (static_cast<int>(temporary_error) == error_code.value()) {
+      return CouchbaseErrorType::TEMPORARY;
+    }
+  }
+  return CouchbaseErrorType::FATAL;
+}
+
+}  // namespace
+
+nonstd::expected<::couchbase::collection, CouchbaseErrorType> CouchbaseClient::getCollection(const CouchbaseCollection& collection) {
+  auto connection_result = establishConnection();
+  if (!connection_result) {
+    return nonstd::make_unexpected(connection_result.error());
+  }
+  std::lock_guard<std::mutex> lock(cluster_mutex_);
+  return cluster_->bucket(collection.bucket_name).scope(collection.scope_name).collection(collection.collection_name);
+}
```

Comment:
I'm not sure if that would be an improvement. I think it's a good practice to keep the scope of locking a mutex as small as possible, to only guard the object (in this case the cluster_) the mutex is meant to synchronize between threads. Even if it would be only 2 additional lines, I wouldn't add it to the scope of the lock guard, and this also looks cleaner this way without passing a lock between the methods.

Reply:
The scope of locking is not different, because it's already held throughout the function, but releasing and locking it again in the middle is unnecessary and probably less efficient.

Reply:
We would need to lock the error check lines after establishConnection that is not needed here, also I think performancewise it would not make a difference, and I thnk passing lock between functions is more unnecessary.

Reply:
Splitting the critical section in half means more context switches, and more cache synchronization or trashing due to different threads accessing the same data. I think it's faster when a single thread can complete its task all at once, without getting interrupted (and cache trashed) by another thread sharing the same data. But I don't remember any references, and it's probably not a huge deal, so I'll yield.